File: packages/go/mattersdk/committee.go
package mattersdk

import (
	cryptorand "crypto/rand"
	"fmt"
	"math/rand/v2"
	"sort"
	"strings"
)

// CommitteeNode is one committee node with the chain-derived data a decryptor needs.
type CommitteeNode struct {
	Index           uint64
	Endpoint        string
	ShareCommitment []byte
}

// DecryptParams is everything needed to recover one secret; the caller supplies
// the chain-derived fields.
type DecryptParams struct {
	SecretID  SecretID
	Epoch     uint32
	BindingID []byte
	Aad       Aad
	Capsule   []byte
	CT        []byte
	SharedA   []byte
	BlockHash [32]byte
	Threshold int
	Nodes     []CommitteeNode
}

// FaultStage says where in the decrypt round trip a node stopped being usable.
type FaultStage string

const (
	// FaultHealth: the /health probe did not answer.
	FaultHealth FaultStage = "health"
	// FaultInactive: /health answered, but the node did not report itself active.
	FaultInactive FaultStage = "inactive"
	// FaultPartialDecrypt: /partial-decrypt did not answer.
	FaultPartialDecrypt FaultStage = "partial-decrypt"
	// FaultEpochMismatch: the node served a different epoch than the caller's
	// state is for, and too few nodes agreed with it to call it a rotation.
	FaultEpochMismatch FaultStage = "epoch-mismatch"
	// FaultProtocolVersion: /health reported a crypto protocol version this
	// SDK does not speak.
	FaultProtocolVersion FaultStage = "protocol-version"
)

// speaksOurProtocol reports whether a node's /health version is compatible. 0
// (field absent) is accepted: each proof's version tag still binds the transcript.
func speaksOurProtocol(reported uint16) bool {
	return reported == 0 || reported == CryptoProtocolVersion()
}

// chooseQuorum picks threshold of available uniformly at random, returned in
// index order (the signed subset's order). Random so no fixed node sees, or can
// deny, every decrypt.
func chooseQuorum(available []CommitteeNode, threshold int, r *rand.Rand) []CommitteeNode {
	shuffled := append([]CommitteeNode(nil), available...)
	r.Shuffle(len(shuffled), func(i, j int) { shuffled[i], shuffled[j] = shuffled[j], shuffled[i] })
	chosen := shuffled[:threshold]
	sort.Slice(chosen, func(i, j int) bool { return chosen[i].Index < chosen[j].Index })
	return chosen
}

// NodeFault is why one committee node did not contribute to a quorum. It never
// carries request material: no signature, auth fields, or partial.
type NodeFault struct {
	Index    uint64
	Endpoint string
	Stage    FaultStage
	// Detail is the underlying reason, already rendered.
	Detail string
}

// DecryptError is a typed decrypt failure; branch on Kind, not on message text.
type DecryptError struct {
	Kind string // "quorum" | "epoch" | "transport" | "crypto"
	Msg  string
	// Faults, for Kind == "quorum", lists which nodes were dropped and why.
	// Empty means the caller supplied too few nodes.
	Faults []NodeFault
}

func (e *DecryptError) Error() string { return e.Msg }

// summarizeFaults renders faults into the message, so a caller that only logs
// err.Error() still gets the reasons.
func summarizeFaults(faults []NodeFault) string {
	if len(faults) == 0 {
		return ""
	}
	parts := make([]string, 0, len(faults))
	for _, f := range faults {
		parts = append(parts, fmt.Sprintf("node %d (%s) %s: %s", f.Index, f.Endpoint, f.Stage, f.Detail))
	}
	return " — " + strings.Join(parts, "; ")
}

// quorumUnavailable builds the error both exhaustion paths return.
func quorumUnavailable(needed, found int, faults []NodeFault) *DecryptError {
	return &DecryptError{
		Kind:   "quorum",
		Msg:    fmt.Sprintf("quorum unavailable: need %d healthy nodes, found %d%s", needed, found, summarizeFaults(faults)),
		Faults: faults,
	}
}

// Decrypt recovers a secret by aggregating a threshold quorum of partial decryptions:
// health-probe -> pick Threshold active nodes at random -> sign once -> query each ->
// verify + aggregate + AEAD-open. The returned plaintext is sensitive.
func Decrypt(transport Transport, signer Signer, p DecryptParams) ([]byte, error) {
	// 1. Health-probe; keep the active nodes and record why others dropped out.
	var active []CommitteeNode
	var faults []NodeFault
	for _, node := range p.Nodes {
		h, err := transport.Health(node.Endpoint)
		switch {
		case err != nil:
			faults = append(faults, NodeFault{node.Index, node.Endpoint, FaultHealth, err.Error()})
		case !speaksOurProtocol(h.CryptoProtocolVersion):
			faults = append(faults, NodeFault{node.Index, node.Endpoint, FaultProtocolVersion,
				fmt.Sprintf("speaks crypto protocol v%d, this SDK speaks v%d", h.CryptoProtocolVersion, CryptoProtocolVersion())})
		case h.Status != "active":
			faults = append(faults, NodeFault{node.Index, node.Endpoint, FaultInactive,
				fmt.Sprintf("status %q, epoch %d, crypto protocol v%d", h.Status, h.Epoch, h.CryptoProtocolVersion)})
		default:
			active = append(active, node)
		}
	}
	if len(active) < p.Threshold {
		return nil, quorumUnavailable(p.Threshold, len(active), faults)
	}

	// 2. Form a quorum. A node that fails or serves a different epoch is dropped
	//    and the subset re-formed, so one node cannot deny the decrypt. A genuine
	//    rotation shows as Threshold nodes agreeing on a new served_epoch. Each
	//    fault removes a node, so the loop terminates.
	available := active
	rotatedVotes := map[uint32]int{}
	r := rand.New(rand.NewChaCha8(randomSeed()))

	for len(available) >= p.Threshold {
		chosen := chooseQuorum(available, p.Threshold, r)
		subset := make([]uint64, len(chosen))
		for i, n := range chosen {
			subset[i] = n.Index
		}

		// 3. Sign per node: the payload binds the node's index, so a signature
		//    cannot be replayed to a peer.
		partials := make([]PartialInput, 0, len(chosen))
		faulty := uint64(0)
		hasFaulty := false

		for _, node := range chosen {
			lambda, err := LagrangeFor(node.Index, subset)
			if err != nil {
				return nil, err
			}
			auth, err := signer.Authorize(p.SecretID, subset, node.Index, p.BlockHash)
			if err != nil {
				return nil, err
			}
			req := PartialDecryptRequest{
				SecretID:      p.SecretID.Hex(),
				Subset:        subset,
				LagrangeCoeff: toHex(lambda),
				Requester:     auth.Requester,
				BlockHash:     toHex(p.BlockHash[:]),
				Signature:     auth.Signature,
				Auth:          auth.Auth,
			}
			resp, err := transport.PartialDecrypt(node.Endpoint, req)
			if err != nil {
				// A per-node fault; the reason distinguishes a refusal after
				// /health from an unreachable node.
				faults = append(faults, NodeFault{node.Index, node.Endpoint, FaultPartialDecrypt, err.Error()})
				faulty, hasFaulty = node.Index, true
				break
			}

			// A different served_epoch from Threshold nodes is a real rotation
			// (fail so the caller refetches); from fewer, a misbehaving node to
			// drop. 0 means an older node serving the current epoch.
			if resp.ServedEpoch != 0 && resp.ServedEpoch != p.Epoch {
				rotatedVotes[resp.ServedEpoch]++
				if rotatedVotes[resp.ServedEpoch] >= p.Threshold {
					return nil, &DecryptError{Kind: "epoch", Msg: fmt.Sprintf(
						"secret served under epoch %d, state supplied for %d; refetch and retry", resp.ServedEpoch, p.Epoch)}
				}
				faults = append(faults, NodeFault{node.Index, node.Endpoint, FaultEpochMismatch,
					fmt.Sprintf("served epoch %d, state supplied for %d", resp.ServedEpoch, p.Epoch)})
				faulty, hasFaulty = node.Index, true
				break
			}

			partial, err := fromHex(resp.Partial)
			if err != nil {
				return nil, &DecryptError{Kind: "transport", Msg: "bad partial hex: " + err.Error()}
			}
			proof, err := fromHex(resp.Proof)
			if err != nil {
				return nil, &DecryptError{Kind: "transport", Msg: "bad proof hex: " + err.Error()}
			}
			partials = append(partials, PartialInput{Point: node.Index, Partial: partial, Proof: proof, Commitment: node.ShareCommitment})
		}

		if hasFaulty {
			remaining := make([]CommitteeNode, 0, len(available)-1)
			for _, n := range available {
				if n.Index != faulty {
					remaining = append(remaining, n)
				}
			}
			available = remaining
			continue
		}
		return openAndWrap(p, partials)
	}

	// Ran out of good nodes. If divergent served_epochs dominated, surface the
	// most common one as a rotation (refetch + retry); otherwise no quorum.
	bestEpoch, bestVotes := uint32(0), 0
	for epoch, votes := range rotatedVotes {
		if votes > bestVotes || (votes == bestVotes && epoch < bestEpoch) {
			bestEpoch, bestVotes = epoch, votes
		}
	}
	if bestVotes > 0 {
		return nil, &DecryptError{Kind: "epoch", Msg: fmt.Sprintf(
			"secret served under epoch %d, state supplied for %d; refetch and retry", bestEpoch, p.Epoch)}
	}
	return nil, quorumUnavailable(p.Threshold, len(available), faults)
}

// openAndWrap verifies, aggregates, and AEAD-opens in the shared core.
func openAndWrap(p DecryptParams, partials []PartialInput) ([]byte, error) {
	pt, err := OpenSecret(p.SharedA, p.Capsule, p.SecretID.Bytes(), p.Epoch, p.BindingID, AadBytes(p.Aad), p.CT, partials)
	if err != nil {
		return nil, &DecryptError{Kind: "crypto", Msg: err.Error()}
	}
	return pt, nil
}

// randomSeed draws a ChaCha8 seed from the OS, so quorum choice cannot be
// predicted from outside the process.
func randomSeed() [32]byte {
	var seed [32]byte
	if _, err := cryptorand.Read(seed[:]); err != nil {
		// crypto/rand does not fail on a supported platform; if it does,
		// nothing else in this process is safe either.
		panic("matter-sdk: crypto/rand: " + err.Error())
	}
	return seed
}

File: bindings/python/python/matter_sdk/committee.py
"""Threshold-decrypt orchestration: form a quorum, sign per node, fan out, open.

Crypto runs in the compiled core. Port of ``packages/typescript-core/src/committee.ts``.
"""

import random
from dataclasses import dataclass
from typing import Dict, List, Optional, Sequence, Union

from ._native import CRYPTO_PROTOCOL_VERSION, lagrange_for, open_secret
from .aad import Aad, aad_bytes
from .errors import DecryptError, NodeFault
from .hexutil import from_hex, secret_id_to_hex, to_hex
from .signer import Signer
from .transport import Transport


@dataclass
class CommitteeNode:
    """One committee node, with the chain-derived data a decryptor needs."""

    index: int
    endpoint: str
    share_commitment: bytes


@dataclass
class DecryptParams:
    """Everything needed to recover one secret (chain-derived fields supplied by you)."""

    secret_id: int
    epoch: int
    binding_id: bytes
    aad: Union[Aad, str, bytes]
    capsule: bytes
    ct: bytes
    shared_a: bytes
    block_hash: bytes
    threshold: int
    nodes: List[CommitteeNode]


def wipe(buf: bytearray) -> None:
    """Zero ``buf`` in place.

    Call it on a recovered plaintext once you are done with it: the core wipes
    its own copy, and this clears the one Python holds.
    """
    buf[:] = bytes(len(buf))


def _speaks_our_protocol(reported: Optional[int]) -> bool:
    """Absent or ``0`` (a node predating the field) is accepted: the version tag on
    every proof still binds the transcript."""
    return not reported or reported == CRYPTO_PROTOCOL_VERSION


_SYSTEM_RANDOM = random.SystemRandom()

#: Auth fields an Ethereum (EIP-712) signer adds to a partial-decrypt request.
_ETHEREUM_AUTH_FIELDS = ("eth_address", "valid_until", "eth_signature")


def choose_quorum(
    available: Sequence[CommitteeNode], threshold: int, rng: random.Random = _SYSTEM_RANDOM
) -> List[CommitteeNode]:
    """Choose ``threshold`` of ``available`` uniformly at random, in index order.

    Random, so no fixed node sees (or can deny) every decrypt. ``rng`` is for
    tests only.
    """
    return sorted(rng.sample(list(available), threshold), key=lambda n: n.index)


def _summarize(faults: Sequence[NodeFault]) -> str:
    if not faults:
        return ""
    return " — " + "; ".join(f"node {f.index} ({f.endpoint}) {f.stage}: {f.detail}" for f in faults)


def _quorum_unavailable(needed: int, found: int, faults: Sequence[NodeFault]) -> DecryptError:
    return DecryptError(
        "quorum",
        f"quorum unavailable: need {needed} healthy nodes, found {found}{_summarize(faults)}",
        faults,
    )


def _rotated(served: int, provided: int) -> DecryptError:
    return DecryptError(
        "epoch", f"secret served under epoch {served}, state supplied for {provided}; refetch and retry"
    )


def decrypt(transport: Transport, signer: Signer, params: DecryptParams) -> bytearray:
    """Recover a secret by aggregating a threshold quorum of partial decryptions.

    A node that fails, or serves a different epoch than ``threshold`` of its
    peers, is dropped and named in :attr:`DecryptError.faults` rather than allowed
    to deny the decrypt. Raises :class:`DecryptError` (``quorum``, ``epoch``, or
    ``crypto``).

    Returns the plaintext as a ``bytearray``: never log it, and :func:`wipe` it
    when done.
    """
    # 1. Health-probe; keep active nodes and record why others dropped out.
    faults: List[NodeFault] = []
    active: List[CommitteeNode] = []
    for node in params.nodes:
        try:
            health = transport.health(node.endpoint)
        except DecryptError as e:
            faults.append(NodeFault(node.index, node.endpoint, "health", str(e)))
            continue
        version = health.get("crypto_protocol_version")
        if not _speaks_our_protocol(version):
            faults.append(
                NodeFault(
                    node.index,
                    node.endpoint,
                    "protocol-version",
                    f"speaks crypto protocol v{version}, this SDK speaks v{CRYPTO_PROTOCOL_VERSION}",
                )
            )
        elif health.get("status") != "active":
            faults.append(
                NodeFault(
                    node.index,
                    node.endpoint,
                    "inactive",
                    f"status {health.get('status')!r}, epoch {health.get('epoch', 0)}, "
                    f"crypto protocol v{version or 0}",
                )
            )
        else:
            active.append(node)

    if len(active) < params.threshold:
        raise _quorum_unavailable(params.threshold, len(active), faults)

    # 2. Assemble a quorum. A faulty node is dropped and the subset re-formed, so
    #    one node cannot deny the decrypt; a real rotation is `threshold` nodes
    #    agreeing on a new served_epoch. Each fault removes a node, so this ends.
    available = active
    rotated_votes: Dict[int, int] = {}

    while len(available) >= params.threshold:
        chosen = choose_quorum(available, params.threshold)
        subset = [n.index for n in chosen]
        partials = []
        faulty: Optional[int] = None

        # 3. Query each chosen node, signing per node so a signature cannot be
        #    replayed to a peer.
        for node in chosen:
            lam = lagrange_for(node.index, subset)
            auth = signer.authorize(params.secret_id, subset, node.index, params.block_hash)
            req = {
                "secret_id": secret_id_to_hex(params.secret_id),
                "subset": subset,
                "lagrange_coeff": to_hex(lam),
                "requester": auth["requester"],
                "block_hash": to_hex(params.block_hash),
                "signature": auth["signature"],
                "auth": auth["auth"],
            }
            # A custom Ethereum-auth signer supplies these; a substrate one omits them.
            for field in _ETHEREUM_AUTH_FIELDS:
                if auth.get(field) is not None:
                    req[field] = auth[field]
            try:
                resp = transport.partial_decrypt(node.endpoint, req)
            except DecryptError as e:
                faults.append(NodeFault(node.index, node.endpoint, "partial-decrypt", str(e)))
                faulty = node.index
                break

            # A different served_epoch: `threshold` agreeing is a real rotation
            # (the caller refetches); fewer is a misbehaving node, dropped.
            # 0/None is an older node serving the current epoch.
            served = resp.get("served_epoch")
            if served and served != params.epoch:
                rotated_votes[served] = rotated_votes.get(served, 0) + 1
                if rotated_votes[served] >= params.threshold:
                    raise _rotated(served, params.epoch)
                faults.append(
                    NodeFault(
                        node.index,
                        node.endpoint,
                        "epoch-mismatch",
                        f"served epoch {served}, state supplied for {params.epoch}",
                    )
                )
                faulty = node.index
                break

            partials.append(
                (node.index, from_hex(resp["partial"]), from_hex(resp["proof"]), bytes(node.share_commitment))
            )

        if faulty is not None:
            available = [n for n in available if n.index != faulty]
            continue

        # 4. Verify + aggregate + AEAD-open in the compiled core.
        try:
            return open_secret(
                params.shared_a,
                params.capsule,
                params.secret_id,
                params.epoch,
                params.binding_id,
                aad_bytes(params.aad),
                params.ct,
                partials,
            )
        except ValueError as e:
            raise DecryptError("crypto", str(e)) from e

    # Out of good nodes: report the most common divergent epoch as a rotation, if any.
    if rotated_votes:
        served = max(rotated_votes, key=lambda epoch: (rotated_votes[epoch], -epoch))
        raise _rotated(served, params.epoch)
    raise _quorum_unavailable(params.threshold, len(available), faults)
